File: templates/python/src/mcp_server/api_client.py
"""Thin client for the upstream API.

This is the only place that knows about base URLs, auth headers, and
timeouts. Tools call typed methods and never see credentials.

Replace the Items methods with calls to the real API you are wrapping.
"""

from __future__ import annotations

from typing import Any

import httpx
from pydantic import BaseModel, Field

from .config import Config


class ApiError(Exception):
    def __init__(self, status: int, message: str, retry_after_seconds: int | None = None) -> None:
        super().__init__(message)
        self.status = status
        self.message = message
        self.retry_after_seconds = retry_after_seconds


class Item(BaseModel):
    id: str
    name: str
    description: str = ""
    created_at: str = Field(alias="createdAt")

    model_config = {"populate_by_name": True}


class ItemPage(BaseModel):
    items: list[Item]
    total: int
    offset: int
    limit: int


class ApiClient:
    def __init__(
        self,
        base_url: str,
        api_key: str = "",
        timeout_ms: int = 15_000,
        transport: httpx.AsyncBaseTransport | None = None,
    ) -> None:
        headers = {"Accept": "application/json"}
        if api_key:
            headers["Authorization"] = f"Bearer {api_key}"
        self._client = httpx.AsyncClient(
            base_url=base_url,
            headers=headers,
            timeout=httpx.Timeout(timeout_ms / 1000),
            transport=transport,
        )

    @classmethod
    def from_config(cls, config: Config, transport: httpx.AsyncBaseTransport | None = None) -> ApiClient:
        return cls(config.api_base_url, config.api_key, config.api_timeout_ms, transport)

    async def aclose(self) -> None:
        await self._client.aclose()

    # ---- Endpoint methods. Replace these for your API. ----------------------

    async def list_items(self, query: str | None, limit: int, offset: int) -> ItemPage:
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if query:
            params["q"] = query
        return ItemPage.model_validate(await self._request("GET", "/items", params=params))

    async def get_item(self, item_id: str) -> Item:
        return Item.model_validate(await self._request("GET", f"/items/{item_id}"))

    async def create_item(self, name: str, description: str | None) -> Item:
        body: dict[str, Any] = {"name": name}
        if description is not None:
            body["description"] = description
        return Item.model_validate(await self._request("POST", "/items", json=body))

    # ---- Transport ----------------------------------------------------------

    async def _request(self, method: str, path: str, **kwargs: Any) -> Any:
        try:
            res = await self._client.request(method, path, **kwargs)
        except httpx.TimeoutException as exc:
            raise ApiError(0, "timed out") from exc
        except httpx.HTTPError as exc:
            raise ApiError(0, "unreachable") from exc

        if res.is_error:
            retry_after = res.headers.get("retry-after")
            raise ApiError(
                res.status_code,
                _extract_error_message(res),
                int(retry_after) if retry_after and retry_after.isdigit() else None,
            )
        if res.status_code == 204 or not res.content:
            return None
        return res.json()


def _extract_error_message(res: httpx.Response) -> str:
    try:
        data = res.json()
    except ValueError:
        return f"HTTP {res.status_code}"
    if isinstance(data, dict):
        err = data.get("error")
        if isinstance(err, str):
            return err
        if isinstance(err, dict) and isinstance(err.get("message"), str):
            return err["message"]
        if isinstance(data.get("message"), str):
            return data["message"]
    return f"HTTP {res.status_code}"

File: templates/typescript/src/api-client.ts
/**
 * Thin client for the upstream API. This is the only place that knows about
 * base URLs, auth headers, and timeouts. Tools call typed methods on it and
 * never see credentials.
 *
 * Replace the Items methods with calls to the real API you are wrapping.
 */
import type { Config } from "./config.js";

export class ApiError extends Error {
  constructor(
    public readonly status: number,
    message: string,
    public readonly retryAfterSeconds?: number,
  ) {
    super(message);
    this.name = "ApiError";
  }
}

export interface Item {
  id: string;
  name: string;
  description: string;
  createdAt: string;
}

export interface ItemPage {
  items: Item[];
  total: number;
  offset: number;
  limit: number;
}

export interface ApiClientOptions {
  baseUrl: string;
  apiKey: string;
  timeoutMs: number;
  /** Injectable for tests. Defaults to global fetch. */
  fetchImpl?: typeof fetch;
}

export class ApiClient {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly timeoutMs: number;
  private readonly fetchImpl: typeof fetch;

  constructor(opts: ApiClientOptions) {
    this.baseUrl = opts.baseUrl;
    this.apiKey = opts.apiKey;
    this.timeoutMs = opts.timeoutMs;
    this.fetchImpl = opts.fetchImpl ?? fetch;
  }

  static fromConfig(config: Config, fetchImpl?: typeof fetch): ApiClient {
    return new ApiClient({
      baseUrl: config.apiBaseUrl,
      apiKey: config.apiKey,
      timeoutMs: config.apiTimeoutMs,
      fetchImpl,
    });
  }

  // ---- Endpoint methods. Replace these for your API. ----------------------

  listItems(params: { query?: string; limit: number; offset: number }): Promise<ItemPage> {
    const search = new URLSearchParams();
    if (params.query) search.set("q", params.query);
    search.set("limit", String(params.limit));
    search.set("offset", String(params.offset));
    return this.request<ItemPage>("GET", `/items?${search.toString()}`);
  }

  getItem(id: string): Promise<Item> {
    return this.request<Item>("GET", `/items/${encodeURIComponent(id)}`);
  }

  createItem(input: { name: string; description?: string }): Promise<Item> {
    return this.request<Item>("POST", "/items", input);
  }

  // ---- Transport ----------------------------------------------------------

  private async request<T>(method: string, path: string, body?: unknown): Promise<T> {
    const headers = new Headers({ Accept: "application/json" });
    if (this.apiKey) headers.set("Authorization", `Bearer ${this.apiKey}`);
    if (body !== undefined) headers.set("Content-Type", "application/json");

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    let res: Response;
    try {
      res = await this.fetchImpl(`${this.baseUrl}${path}`, {
        method,
        headers,
        body: body === undefined ? undefined : JSON.stringify(body),
        signal: controller.signal,
      });
    } catch (err) {
      const reason = err instanceof Error && err.name === "AbortError" ? "timed out" : "unreachable";
      throw new ApiError(0, `Upstream API ${reason}`);
    } finally {
      clearTimeout(timer);
    }

    if (!res.ok) {
      const retryAfter = Number(res.headers.get("retry-after"));
      throw new ApiError(
        res.status,
        await extractErrorMessage(res),
        Number.isFinite(retryAfter) && retryAfter > 0 ? retryAfter : undefined,
      );
    }
    if (res.status === 204) return undefined as T;
    return (await res.json()) as T;
  }
}

async function extractErrorMessage(res: Response): Promise<string> {
  try {
    const data = (await res.json()) as { error?: { message?: string } | string; message?: string };
    if (typeof data.error === "string") return data.error;
    if (data.error?.message) return data.error.message;
    if (data.message) return data.message;
  } catch {
    // not JSON
  }
  return `HTTP ${res.status}`;
}
